### ISSUE TITLE
Problems with Local adapter and r/o mount filesystem

### ISSUE THREAD

**Issue by maschmann** [thephpleague/flysystem#518]
Hi,

I'm using flysystem to abstract several storage engines where we keep our images. So far, so good -On test/live systems we're using a r/o NFS mount. I'm also using the symfony2 integration bundle (oneup/flysystem) to make the library available in my symfony2 application.
I'm now facing the problem that during the cache:clear command, the service container and therefore the flysystem adapter are initialized. This leads to a warning, causing my application to crash, because flysystem local adapter tries to write to the r/o filesystem, which fails.

Stacktrace:
```php
  [Symfony\Component\Debug\Exception\ContextErrorException]  
  Warning: mkdir(): Permission denied 

Exception trace:
 () at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/league/flysystem/src/Adapter/Local.php:55
 Symfony\Component\Debug\ErrorHandler->handleError() at n/a:n/a
 mkdir() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/league/flysystem/src/Adapter/Local.php:55
 League\Flysystem\Adapter\Local->ensureDirectory() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/league/flysystem/src/Adapter/Local.php:35
 League\Flysystem\Adapter\Local->__construct() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/cache/stag_/ap_StageDebugProjectContainer.php:2606
 ap_StageDebugProjectContainer->getOneupFlysystem_MhImageFilesystemService() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/bootstrap.php.cache:2110
 Symfony\Component\DependencyInjection\Container->get() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/cache/stag_/ap_StageDebugProjectContainer.php:2622
 ap_StageDebugProjectContainer->getOneupFlysystem_MountManagerService() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/bootstrap.php.cache:2110
 Symfony\Component\DependencyInjection\Container->get() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/cache/stag_/ap_StageDebugProjectContainer.php:489
 ap_StageDebugProjectContainer->getCore_Content_MediaService() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/bootstrap.php.cache:2110
 Symfony\Component\DependencyInjection\Container->get() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/cache/stag_/ap_StageDebugProjectContainer.php:779
 ap_StageDebugProjectContainer->getCore_Twig_MpsMediaExtensionService() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/bootstrap.php.cache:2110
 Symfony\Component\DependencyInjection\Container->get() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/cache/stag_/ap_StageDebugProjectContainer.php:4105
 ap_StageDebugProjectContainer->getTwigService() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/bootstrap.php.cache:2110
 Symfony\Component\DependencyInjection\Container->get() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Bundle/TwigBundle/CacheWarmer/TemplateCacheCacheWarmer.php:54
 Symfony\Bundle\TwigBundle\CacheWarmer\TemplateCacheCacheWarmer->warmUp() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:48
 Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php:138
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->warmup() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php:90
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->execute() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:257
 Symfony\Component\Console\Command\Command->run() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:882
 Symfony\Component\Console\Application->doRunCommand() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:195
 Symfony\Component\Console\Application->doRun() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:96
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:126
 Symfony\Component\Console\Application->run() at /var/www/vhosts/mps_frontend/htdocs/releases/20150630111910_master/app/console:27


cache:clear [--no-warmup] [--no-optional-warmers]
```

Is there a way to suppress the mkdir?

**Comment by frankdejonge** [comment]
@maschmann isn't there a way to prevent the filesystem from being initialised in the first place?

**Comment by maschmann** [comment]
@frankdejonge afaik only if I change the way the services in the flysystem bundle are initialized (lazy). But I'm not sure if that works, either.

**Comment by frankdejonge** [comment]
@maschmann the mkdir call only happens if you instantiate a local filesystem adapter for a root path that doesn't exist, this to prevent having to check for it's existence on every call. So it's the lesser of two evils. I do not think changing this behaviour is in flysystem's best interest.

**Comment by maschmann** [comment]
@frankdejonge I've had a look into the code and I think the problem is not the readonly filesystem, but that it's a mounted (NFS) directory. For test/production systems we've got shared files (media, images, ...) on a NFS storage that we mount on all frontend machines. r/o because data is written by only one master instance of the CMS that's providing the data for the frontends.
I'm not sure if that's solvable, since is_dir is php spl -.-

**Comment by Markcial** [comment]
do we really need to check for filesystem capabilities on the construct for Local adapter?

we are having a little trouble with that too, we have a filesystem that does writes for command calls but when the service is loaded from within the web dependency injection throws exceptions because the user is not able to do the same operations as the CLI user.

If a Filesystem adapter is loaded but not meant to be used there is no need to check if the adapter is capable of some operations, the check must be done before any R/W operation but on construct.

Would be nice to have the Local adapter loaded so it doesn't checks for FS operationality until it is requested to do so.

Thanks

**Comment by frankdejonge** [comment]
@maschmann this entire problem can me mitigated by defining the flysystem service as lazy.

**Comment by jfoliveira** [comment]
I totally agree we don't need to change this behavior in flysystem, but it would help a lot if the error message was user friendly.

`Permission denied` is too generic in this context, as many times this issue only means that directory that is being ensured to exist by `ensureDirectory` method simply can't exist. That happens, for instance, if the value of `$root` points to a directory that is not accessible in the current file system.

How to improve it? By simply wrapping the existing code with a `try..finally` block the error message would display the directory that was about to be created, therefore allowing the developer to understand the reason behind the error and make the necessary adjustments:

```php
    protected function ensureDirectory($root)
    {
        if ( ! is_dir($root)) {
            $umask = umask(0);
            // dd($root, $this->permissionMap['dir']['public']);
            try {
                mkdir($root, $this->permissionMap['dir']['public'], true);
                umask($umask);
            } finally {
                if ( ! is_dir($root)) {
                    throw new Exception(sprintf('Impossible to create the root directory "%s".', $root));
                }
            }

        }

        return realpath($root);
    }
```

So now, with the above change, instead of `Permission denied` error one would see the error message:
```
Impossible to create the root directory "/some/invalid/dir".
```

Hope that helps other people struggling with this - as I did struggle today.

**Comment by frankdejonge** [comment]
@jfoliveira I like your suggestion 👍

**Comment by chapterjason** [comment]
@jfoliveira @frankdejonge Change of message is good, but still very confusing. I assume that the change https://github.com/thephpleague/flysystem-bundle/commit/c8e55b6a30792ee721d62847e08e76dcd74c3371 already set the services to lazy by default. The error still occurs in my case.

After debugging, I noticed that the function gets called with directory `/app/var/storage/default/cover` but the actual root directory is `/app/var/storage/default`.
